### ISSUE TITLE
another error in WQ_gui_branch

### ISSUE THREAD

**Issue by bumpersp** [mkleinha/sqlite_interface#5]
I'm now getting this error when I tried to push data to the habitat table. Error in if: argument is of length zero. 

This kind of debugging is still a bit beyond me as I don't actually know what i'm doing yet-I'm just good at copy and paste. 

I'll get there.

**Comment by mkleinha** [comment]
Debugging shiny code is definitely a pain compared to normal scripts. I think the problem here is that in a selectInput() where the parameter 'multiple' is set to TRUE, the empty value returned is a vector of length 0 instead of NA. For those inputs where we want to accept multiple values, we might have to check for good values using if(length('input_name') < 1) instead of if(is.na('input_name')). On that note, I think it might be worth deliberately assessing which inputs actually need to accept multiple values and setting that parameter accordingly. 
On the debugging front, are you looking at the output in the R console in the Rstudio window? The errors output there are much more descriptive (including backtraces and line numbers) than the output printed to the shiny app interface.
It's awesome you put this together as fast as you did by the way.

**Comment by bumpersp** [comment]
Thanks, Max. Yeah I was looking at the console for expanded error messages but for some reason I'm not seeing line numbers for these errors. This is the extent of this error in my console "Warning: Error in if: argument is of length zero
  [No stack trace available]". i usually am able to detect things like missing paranetheses, for example, as problems with the red x circles that show up on the left had side of the R script but that's not showing up for these errors. 

Thanks for your help with this. To be clear about your comment, are you suggesting that for example this line: 
#need to select substrate types
  if(is.na(Substrate)){
    msg <- paste0(msg, "No substrate selected.<br/>")
  }
Should perhaps be:
if(length(Substrate)<1){
msg <- paste0(msg, "No substrate selected.<br/>")
  }
??

THanks for noticing the multiple=True. I didn't mean to have that set for as meaning inputs as i did. I think i just copied and pasted that line of code and forgot to delete that for inputs like "Collection_Type" that don't need multiple.

**Comment by bumpersp** [comment]
Following up with this error code that i've yet to debug. 

Warning: Error in vapply: values must be length 1,
 but FUN(X[[24]]) result is length 0 @mkleinha

**Comment by mkleinha** [comment]
In response to your question from yesterday, that's exactly what I meant for the if() statements.

I think the new error means that sqlInterpolate() can't find one of the parameters it's expecting. There might be another small typo between those parameters, the ones in the query string, and the parameters of the update_hab() function. I don't really want to take the time to go through and check them all to see if my suspicion is correct. The number of parameters to this function has really ballooned now that it has been expanded to accept all the relevant water quality fields. It makes me think we should try to simplify this some since it's getting hard for me to track all the parameters. The fewer times we write them out the less likely it is we'll make typos. I think in hindsight, the extra step of including the sqlInterpolate() function and separate construction of the SQL query might be unnecessary. My data entry example was adapted from the mitigation bank script, which was meant to define a public interface. The sqlInterpolate() call was included as a guard against SQL injection attacks on the database. Since the interface we're working on now will probably only be used internally by RBC, I see no reason to worry about this brand of security. I'll try altering the new example script to eliminate this function.

**Comment by bumpersp** [comment]
Thanks. That was my assumption but as you say the parameters have ballooned and it's difficult to find typos in all of the lists vs the table itself. I'll keep going over it. But yes, simplifiying that a bit seems like a plus. Do you envision eveuntlay having the 3 seperate pieces of the code being seperate? For instance the Server section being it's on script that gets sourced or should this all be one long file?

**Comment by mkleinha** [comment]
I would be fine with splitting into different files. I mostly kept everything in one file so far to coddle my limited version control skills.

**Comment by mkleinha** [comment]
I added an updated data_entry_example.R file to the database directory on the RBC drive. This one includes querying and editing of database records. The update_hab() function also uses two other methods to add records to the database. I thought both would be simpler than the sqlInterpolate() function but I think the first might actually be fiddlier. Let me know what you think.

**Comment by bumpersp** [comment]
Cool. I'll look this over and do my best to understand the differences. I may need you to walk me through that though. On first glance, the entry form works well. I'm not exactly sure how the edit record works. When I query a record i can't seem to change the value of any parameter.

**Comment by mkleinha** [comment]
The new example script should work now. I had to go back to sqlInterpolate(), unfortunately.

**Comment by bumpersp** [comment]
Yep. That seems to work for me now.

**Comment by bumpersp** [comment]
But the "INSERT INTO HABITAT" portion is still no longer needed? The new data is submitted via the dbWritetable bit with "vals"?

**Comment by mkleinha** [comment]
Right. dbWriteTable() works for adding a record, but I couldn't find a better way than sqlInterpolate() for constructing an UPDATE statement.

**Comment by bumpersp** [comment]
quick quesiton. in your new example code you have Date and date. As far as I can tell 'date' is not defined until the UI as the inputID for the Date window. Then date is used for the input$. Should this all actually be 'Date'. I"m not able to tell where/how these might be different.

**Comment by mkleinha** [comment]
They are different in that 'date' is the ID of the user input, and 'Date' is a parameter of the function update_hab() and the name of the database field. I can get away with this because functions don't care about the names of the parameters they receive, only the values, as long as the parameters are specified in the correct order. These should probably all be the same, or capitalization should consistently separate user inputs from function parameters. I think the former is probably simpler. Sorry about the inconsistencies. Sloppy coding on my part.

**Comment by bumpersp** [comment]
Gotcha okay. Thanks for clarifying that.  And lastly in the last bit toward the bottom with updateTextinput....why are there 2 rows for observers in both sets of this code?

**Comment by mkleinha** [comment]
Well that's a good question, Phillip. I think the reason is that was the one I copied and pasted to make the other lines for that section. It didn't throw an error so I didn't catch the duplication. Woops. sloppy

**Comment by bumpersp** [comment]
That what my assumption-but since it wasn't throwing an error I wasn't sure. 

almost beer 30..I'll stop "bug"ging you....

**Comment by bumpersp** [comment]
Okay,  I hate to do this but I really can't figure out the issue with the update_hab function. Now getting this error when submitting data using the updated code: Warning: Error in data.frame: arguments imply differing number of rows: 1, 0

I've stared at the spellings, everything is in the same order as the table. Gone line by line comparing to your example code.... I can't find it. If you have anytime to troubleshoot this that would be great---but don't spend too much time on it.

**Comment by bumpersp** [comment]
well i'm not exaclty sure what I did, but now this script seems to work. Thanks!

**Comment by mkleinha** [comment]
Alright. Sorry I didn't get at it sooner. I love it when code fixes itself.

**Comment by bumpersp** [comment]
no sweat at all. I had given up though. For some reason tried it again and now I have 4 rows of data submitted....go figure!

**Comment by bumpersp** [comment]
of course....now after 5 entries, it starts throwing the error code again.

**Comment by bumpersp** [comment]
Just fyi, I think I figured it out. I had delete the 
msg <- ""
  if(length(Substrate)<1){
    msg <- paste0(msg, "No substrate selected.")
  }  
code after adjusting for your new query/edit code and that does indeed present a problem. It was not letting me submit data when parameters that were set to multiple = TRUE were empty.
